maro/simulator/scenarios/data_center/business_engine.py

```diff
@@ -0,0 +1,241 @@
+        self._success_requirements: int = 0
+        self._failed_requirements: int = 0
+        self._total_latency: Latency = Latency()
+
+        # Load configurations.
+        self._load_configs()
+        self._register_events()
+
+        # PM initialize.
+        self._init_machines()
+
+        # VM initialize.
+        self._vm: dict = {}
+
+        self._tick: int = 0
+        self._delay_duration: int = self._conf["delay_duration"]
+
+    def _load_configs(self):
+        """Load configurations."""
+
+        with open(os.path.join(self._config_path, "config.yml")) as fp:
+            self._conf = safe_load(fp)
+
+    def _init_machines(self):
+        """Initialize the physical machines based on the config setting. The PM id starts from 0."""
+
+        self._machines: List[PhysicalMachine] = [
+            PhysicalMachine(
+                id=i,
+                cap_cpu=self._conf["pm_cap_cpu"],
+                cap_mem=self._conf["pm_cap_mem"]
+            ) for i in range(self._conf["pm_amount"])
+        ]
+
+    def step(self, tick: int):
+        """Push business to next step.
+
+        Args:
+            tick (int): Current tick to process.
+        """
+        self._tick = tick
+
+        # Update all live VMs memory utilization.
+        self._update_vm_util()
+        # Update all PM memory utilization.
+        self._update_pm_util()
+        # TODO
+        # Generate VM requirement events from data file.
+        # It might be implemented by a for loop to process VMs in each tick.
+        vm_required_evt = self._event_buffer.gen_cascade_event(tick, Events.REQUIREMENTS, payload=None)
+        self._event_buffer.insert_event(vm_required_evt)
+
+    def get_metrics(self) -> dict:
+        """Get current enviornment metrics information.
+
+        Returns:
+            dict: Metrics information.
+        """
+
+        return DocableDict(
+            metrics_desc,
+            energy_consumption=self._energy_consumption,
+            success_requirements=self._success_requirements,
+            failed_requirements=self._failed_requirements,
+            total_latency=self._total_latency
+        )
+
+    def _register_events(self):
+        # Register our own events and their callback handlers.
+        self._event_buffer.register_event_handler(Events.REQUIREMENTS, self._on_vm_required)
+        self._event_buffer.register_event_handler(Events.FINISHED, self._on_vm_finished)
+
+        # Generate decision event.
+        self._event_buffer.register_event_handler(MaroEvents.TAKE_ACTION, self._on_action_received)
+
+    def _update_vm_util(self):
+        """Update all live VMs memory utilization.
+        The length of VMs utilization series could be difference among all VMs,
+        because index 0 represents the VM's CPU utilization at the tick it starts.
+        """
+
+        for _, vm in self._vm.items():
+            vm.util_cpu = vm.util_series[self._tick - vm.start_tick]
```

Comment:
calculate the offset inside vm, not here.

like:

``vm.util_cpu = vm.get_util(env_tick=self.tick)``

or use some other naming